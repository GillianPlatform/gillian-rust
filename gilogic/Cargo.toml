[package]
name = "gilogic"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
gilogic_proc = { path = "../gilogic_proc" }
gilogic_proc_dummy = { path = "../gilogic_proc_dummy" }