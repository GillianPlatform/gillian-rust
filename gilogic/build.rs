fn main() {
    println!("cargo::rustc-check-cfg=cfg(gillian)");
}
