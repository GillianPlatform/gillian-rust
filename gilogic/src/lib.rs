#![allow(internal_features)]
#![feature(rustc_attrs)]
#![feature(never_type)]
#![feature(register_tool)]
#![feature(ptr_internals)]
#![feature(allocator_api)]
#![feature(tuple_trait, unboxed_closures)]
#![feature(stmt_expr_attributes, proc_macro_hygiene)]
#![register_tool(gillian)]

mod tys;

pub use tys::RustAssertion;

pub mod macros {
    pub use gilogic_proc::{
        assertion, assertion_test, borrow, close_borrow, extract_lemma, lemma, open_borrow,
        predicate, show_safety, specification,
    };
    pub mod prophecies {
        pub use gilogic_proc::with_freeze_lemma_for_mutref_pcy as with_freeze_lemma_for_mutref;
    }
    pub mod no_prophecies {
        pub use gilogic_proc::with_freeze_lemma_for_mutref_no_pcy as with_freeze_lemma_for_mutref;
    }
}

mod seq;
pub use seq::Seq;

mod ownable;
pub use ownable::Ownable;
pub mod alloc;
pub mod iterated;
pub mod prophecies;

pub mod symex_ctrl {

    #[gillian::builtin]
    #[gillian::no_translate]
    #[rustc_diagnostic_item = "gillian::symex_ctrl::symex_next"]
    #[inline(always)]
    pub fn symex_next() {}
}

#[path = "stubs.rs"]
pub mod __stubs;
