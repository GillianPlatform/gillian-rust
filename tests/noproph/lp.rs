//@check-pass
extern crate gilogic;

use gilogic::{macros::*, Ownable};

struct LP<T> {
    x: *mut N<T>,
    y: *mut N<T>,
}

struct N<T> {
    v: T,
    o: *mut N<T>,
}

#[predicate]
fn lp<T: Ownable>(lp: In<LP<T>>, x: *mut N<T>, y: *mut N<T>) {
    assertion!(|v_x, v_y|
        (lp == LP { x, y }) *
        (x -> N { v: v_x, o: y }) *
        (y -> N { v: v_y, o: x }) *
        v_x.own() * v_y.own()
    )
}

#[extract_lemma(
    forall x, y.
    from { lp_ref_mut_xy(p, (x, y)) }
    extract { <&mut T as Ownable>::own(&mut (*x).v) }
)]
fn extract_x<'a, T: Ownable>(p: &'a mut LP<T>);

#[extract_lemma(
    forall x, y.
    from { lp_ref_mut_xy(p, (x, y)) }
    extract { <&mut T as Ownable>::own(&mut (*y).v) }
)]
fn extract_y<'a, T: Ownable>(p: &'a mut LP<T>);

#[with_freeze_lemma(
    lemma_name = freeze_xy,
    predicate_name = lp_ref_mut_xy,
    frozen_variables = [x, y],
)]
impl<T: Ownable> Ownable for LP<T> {
    #[predicate]
    fn own(self) {
        assertion!(|x: *mut N<T>, y: *mut N<T>| lp(self, x, y))
    }
}

impl<T: Ownable> LP<T> {
    #[show_safety]
    fn new(x: T, y: T) -> Self {
        let null: *mut N<T> = std::ptr::null_mut();

        let xbox = Box::new(N { v: x, o: null });
        let ybox = Box::new(N { v: y, o: null });

        let xptr = Box::leak(xbox) as *mut N<T>;
        let yptr = Box::leak(ybox) as *mut N<T>;

        unsafe {
            (*yptr).o = xptr;
            (*xptr).o = yptr;
        }
        LP { x: xptr, y: yptr }
    }

    #[show_safety]
    fn assign_first(&mut self, x: T) {
        unsafe {
            (*self.x).v = x;
        }
    }

    #[show_safety]
    fn first_mut<'a>(&'a mut self) -> &'a mut T {
        unsafe {
            freeze_xy(self);
            let ret = &mut (*self.x).v;
            extract_x(self);
            ret
        }
    }

    #[show_safety]
    fn second_mut<'a>(&'a mut self) -> &'a mut T {
        unsafe {
            freeze_xy(self);
            let ret = &mut (*self.y).v;
            extract_y(self);
            ret
        }
    }
}
