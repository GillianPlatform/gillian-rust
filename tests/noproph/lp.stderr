[TIMESTAMP WARN  lib_rtg::metadata] could not read metadata for crate `"std"`: Os { code: 2, kind: NotFound, message: "No such file or directory" }
[TIMESTAMP WARN  lib_rtg::codegen::block] Not handling drop properly yet: _6
[TIMESTAMP WARN  lib_rtg::codegen::block] Not handling drop properly yet: _10
[TIMESTAMP WARN  lib_rtg::codegen::block] Not handling drop properly yet: _8
[TIMESTAMP WARN  lib_rtg::codegen::block] Not handling drop properly yet: _4
[TIMESTAMP WARN  lib_rtg::codegen::block] Not handling drop properly yet: _2
[TIMESTAMP WARN  lib_rtg::codegen::block] Not handling drop properly yet: _1
[TIMESTAMP WARN  lib_rtg::codegen::block] Not handling drop properly yet: ((*((*_1).0: *mut N<T>)).0: T)
[TIMESTAMP WARN  lib_rtg::codegen::block] Not handling drop properly yet: _3
[TIMESTAMP WARN  lib_rtg::codegen::block] Not handling drop properly yet: _2
warning: unused variable: `p`
  --> ../tests/noproph/lp.rs:31:30
   |
31 | fn extract_x<'a, T: Ownable>(p: &'a mut LP<T>);
   |                              ^ help: if this is intentional, prefix it with an underscore: `_p`
   |
   = note: `#[warn(unused_variables)]` on by default

warning: unused variable: `p`
  --> ../tests/noproph/lp.rs:38:30
   |
38 | fn extract_y<'a, T: Ownable>(p: &'a mut LP<T>);
   |                              ^ help: if this is intentional, prefix it with an underscore: `_p`

warning: fields `x` and `y` are never read
 --> ../tests/noproph/lp.rs:7:5
  |
6 | struct LP<T> {
  |        -- fields in this struct
7 |     x: *mut N<T>,
  |     ^
8 |     y: *mut N<T>,
  |     ^
  |
  = note: `#[warn(dead_code)]` on by default

warning: fields `v` and `o` are never read
  --> ../tests/noproph/lp.rs:12:5
   |
11 | struct N<T> {
   |        - fields in this struct
12 |     v: T,
   |     ^
13 |     o: *mut N<T>,
   |     ^

warning: function `extract_x___proof` should have a snake case name
  --> ../tests/noproph/lp.rs:31:4
   |
31 | fn extract_x<'a, T: Ownable>(p: &'a mut LP<T>);
   |    ^^^^^^^^^ help: convert the identifier to snake case: `extract_x_proof`
   |
   = note: `#[warn(non_snake_case)]` on by default

warning: function `extract_y___proof` should have a snake case name
  --> ../tests/noproph/lp.rs:38:4
   |
38 | fn extract_y<'a, T: Ownable>(p: &'a mut LP<T>);
   |    ^^^^^^^^^ help: convert the identifier to snake case: `extract_y_proof`

warning: variable `THIS` should have a snake case name
  --> ../tests/noproph/lp.rs:47:12
   |
47 |     fn own(self) {
   |            ^^^^ help: convert the identifier to snake case: `this`

warning: 7 warnings emitted

[TIMESTAMP INFO  rust_to_gil] Compiled!
